pkg/apis/resource/types.go

```diff
@@ -129,106 +644,142 @@ type ResourceClaimStatus struct {
+const AllocationResultsMaxSize = 32
 
-// StructuredResourceHandle is the in-tree representation of the allocation result.
-type StructuredResourceHandle struct {
-	// VendorClassParameters are the per-claim configuration parameters
-	// from the resource class at the time that the claim was allocated.
-	VendorClassParameters runtime.Object
+// DeviceRequestAllocationResult contains the allocation result for one request.
+type DeviceRequestAllocationResult struct {
+	// Request is the name of the request in the claim which caused this
+	// device to be allocated. Multiple devices may have been allocated
+	// per request.
+	//
+	// +required
+	Request string
 
-	// VendorClaimParameters are the per-claim configuration parameters
-	// from the resource claim parameters at the time that the claim was
-	// allocated.
-	VendorClaimParameters runtime.Object
+	// Driver specifies the name of the DRA driver whose kubelet
+	// plugin should be invoked to process the allocation once the claim is
+	// needed on a node.
+	//
+	// Must be a DNS subdomain and should end with a DNS domain owned by the
+	// vendor of the driver.
+	//
+	// +required
+	Driver string
 
-	// NodeName is the name of the node providing the necessary resources
-	// if the resources are local to a node.
-	NodeName string
+	// This name together with the driver name and the device name field
+	// identify which device was allocated (`<driver name>/<pool name>/<device name>`).
+	//
+	// Must not be longer than 253 characters and may contain one or more
+	// DNS sub-domains separated by slashes.
+	//
+	// +required
+	Pool string
 
-	// Results lists all allocated driver resources.
-	Results []DriverAllocationResult
+	// Device references one device instance via its name in the driver's
+	// resource pool. It must be a DNS label.
+	//
+	// +required
+	Device string
 }
 
-// DriverAllocationResult contains vendor parameters and the allocation result for
-// one request.
-type DriverAllocationResult struct {
-	// VendorRequestParameters are the per-request configuration parameters
-	// from the time that the claim was allocated.
-	VendorRequestParameters runtime.Object
+// DeviceAllocationConfiguration gets embedded in an AllocationResult.
+type DeviceAllocationConfiguration struct {
+	// Source records whether the configuration comes from a class and thus
+	// is not something that a normal user would have been able to set
+	// or from a claim.
+	//
+	// +required
+	Source AllocationConfigSource
 
-	AllocationResultModel
-}
+	// Requests lists the names of requests where the configuration applies.
+	// If empty, its applies to all requests.
+	//
+	// +optional
+	// +listType=atomic
+	Requests []string
 
-// AllocationResultModel must have one and only one field set.
-type AllocationResultModel struct {
-	// NamedResources describes the allocation result when using the named resources model.
-	NamedResources *NamedResourcesAllocationResult
+	DeviceConfiguration // inline
```

Comment:
I guess this is why we do the inlining -- so that we can reuse the same struct in two places? It's probably worth a comment at each usage site to this effect to avoid confusion / questions.

Reply:
I had such meta-comments in the `type.go` and Tim told me to remove them. I'll keep them in the KEP as commentary on the API. I'm aware that this won't help someone reading only the `types.go`, in particular now that the KEP is stale :cry:

Reply:
Something like:

```
This is defined as a distinct type because it is going to be reused in an upcoming release.
"""

...isn't egregious, but it shows up in the client (meh) and in docs (bad), unless there's some trick to making it not show up?

Reply:
I did some experimenting.

It looks like comments that are not directly attached to types or fields do not get propagated into docs, so putting such comments at the end of a type or "detached" should be OK?

It also looks like any comment with a `// +<anything>` form will be filtered out, so perhaps we can invent another convention:

```
// This field is blah blah blah
// +optional
// +// this is an implementation note
// +// these will not show up in docs (except godoc)
```

Reply:
I had such detached comments earlier. For example, all of the "future extension" commands where at the end of a struct after the current fields and thus ignored by API docs and Go docs. Only a developer reading the types.go would see them.

I still have one "meta comment" left, the one about the `Value` suffix in `DeviceAttribute`. That appears before the fields, but because it's detached, it's also ignored.

Embedding with `+//` may be useful when a detached comment is "too far away".

Anyway, adding back such comments is currently lower priority for me. Perhaps when things are more settled...

Reply:
Yeah, I mistakenly thought those would propagate into the docs, but they do not (though I am not sure if that's intentional or accidental).  If you want to add such things back, I am cautiously OK with it as long as it's not phrased as a commitment :)